sys/net/network_layer/fib/fib.c

```diff
@@ -609,6 +691,712 @@ int fib_get_num_used_entries(void)
+        }
+        elt_repl->address = add;
+    }
+
+    mutex_unlock(&mtx_access);
+    return 0;
+}
+
+int fib_sr_entry_get_address(fib_sr_t *fib_sr, fib_sr_entry_t *sr_entry,
+                             uint8_t *addr, size_t *addr_size)
+{
+    timex_t now;
+    vtimer_now(&now);
+
+    mutex_lock(&mtx_access);
+    if (fib_sr == NULL) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    if (fib_sr_check_lifetime(fib_sr, &now) == -ENOENT) {
+        mutex_unlock(&mtx_access);
+        return -ENOENT;
+    }
+
+    fib_sr_entry_t *elt;
+    LL_FOREACH(fib_sr->sr_path, elt) {
+        if (elt == sr_entry) {
+            if (universal_address_get_address(elt->address, addr, addr_size) != NULL) {
+                mutex_unlock(&mtx_access);
+                return 0;
+            }
+            else {
+                mutex_unlock(&mtx_access);
+                return -ENOMEM;
+            }
+        }
+    }
+    mutex_unlock(&mtx_access);
+    return -ENOENT;
+}
+
+int fib_sr_get_route(uint8_t *dst, size_t dst_size, kernel_pid_t *sr_iface_id,
+                     uint32_t *sr_flags,
+                     uint8_t *addr_list, size_t *addr_list_elements, size_t *element_size,
+                     bool reverse)
+{
+    timex_t now;
+    vtimer_now(&now);
+
+    mutex_lock(&mtx_access);
+
+    if ((dst == NULL) || (sr_iface_id == NULL) || (sr_flags == NULL)
+        || (addr_list == NULL) || (addr_list_elements == NULL)
+        || (element_size == NULL)) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    fib_sr_t *hit = NULL;
+    fib_sr_t *tmp_hit = NULL;
+    int check_free_entry = -1;
+    /* Case 1 - check if we know a direct route */
+    for (size_t i = 0; i < FIB_MAX_FIB_SR; ++i) {
+        if (fib_sr_check_lifetime(&fib_sr_table[i], &now) == -ENOENT) {
+            /* expired, so skip this sr and remember its position */
+            check_free_entry = i;
+            continue;
+        }
+
+        size_t addr_size_match = dst_size << 3;
+        if (universal_address_compare(fib_sr_table[i].sr_dest->address,
+                                      dst, &addr_size_match) == 0) {
+            if (*sr_flags == fib_sr_table[i].sr_flags) {
+                /* found a perfect matching sr, no need to search further */
+                hit = &fib_sr_table[i];
+                tmp_hit = NULL;
+                if (check_free_entry == -1) {
+                    check_free_entry = i;
+                }
+                break;
+            }
+            else {
+                /* found a sr to the destination but with different flags,
+                 * maybe we find a better one.
+                 */
+                tmp_hit = &fib_sr_table[i];
+            }
+        }
+    }
+
+    if (hit == NULL) {
+        /* we didn't find a perfect sr, but one with distinct flags */
+        hit = tmp_hit;
+    }
+
+    /* Case 2 - if no hit is found check if there is a matching entry in one sr_path
```

Comment:
Just to check if I understand this correctly:
If this sr exist: `a->b->c->d`
But I query a sr for the destination `c`, then this function will create the new sr: `a->b->c`,
so that the table holds now `a->b->c->d` and `a->b->c`?